File: src/DatabaseManager/Table Edits/CellEdit.cs
/*
* NOTICE:
* The U.S. Army Corps of Engineers, Risk Management Center (USACE-RMC) makes no guarantees about
* the results, or appropriateness of outputs, obtained from this software.
*
* LIST OF CONDITIONS:
* Redistribution and use in source and binary forms, with or without modification, are permitted
* provided that the following conditions are met:
* ● Redistributions of source code must retain the above notice, this list of conditions, and the
* following disclaimer.
* ● Redistributions in binary form must reproduce the above notice, this list of conditions, and
* the following disclaimer in the documentation and/or other materials provided with the distribution.
* ● The names of the U.S. Government, the U.S. Army Corps of Engineers, the Institute for Water
* Resources, or the Risk Management Center may not be used to endorse or promote products derived
* from this software without specific prior written permission. Nor may the names of its contributors
* be used to endorse or promote products derived from this software without specific prior
* written permission.
*
* DISCLAIMER:
* THIS SOFTWARE IS PROVIDED BY THE U.S. ARMY CORPS OF ENGINEERS RISK MANAGEMENT CENTER
* (USACE-RMC) "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
* THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
* DISCLAIMED. IN NO EVENT SHALL USACE-RMC BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
* SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
* PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
* INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
* LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
* THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

using System.Collections.Generic;

namespace DatabaseManager
{
    /// <summary>
    /// Represents an edit to a single cell in a table, storing row index, column index, and the edited value.
    /// </summary>
    public class CellEdit : TableEdit
    {
        private int _rowIndex;
        private int _columnIndex;
        private readonly object _editValue;

        /// <summary>
        /// Gets the row index of the edited cell. 
        /// </summary>
        public int RowIndex
        {
            get
            {
                return _rowIndex;
            }
        }

        /// <summary>
        /// Gets the column index of the edited cell.
        /// </summary>
        public int ColumnIndex
        {
            get
            {
                return _columnIndex;
            }
        }

        /// <summary>
        /// Gets the new value assigned to the cell.
        /// </summary>
        public object Value
        {
            get
            {
                return _editValue;
            }
        }

        /// <summary>
        /// Initializes a new instance of the <see cref="CellEdit"/> class with specified row, column, and value.
        /// </summary>
        /// <param name="indexOfRow">The row index of the edited cell.</param>
        /// <param name="indexOfColumn">The column index of the edited cell.</param>
        /// <param name="editValue">The new value for the cell. </param>
        public CellEdit(int indexOfRow, int indexOfColumn, object editValue)
        {
            _rowIndex = indexOfRow;
            _columnIndex = indexOfColumn;
            _editValue = editValue;
        }

        /// <summary>
        /// Returns the edited cell if its column matches the specified index.
        /// </summary>
        /// <param name="indexOfColumn">The column index to check.</param>
        /// <returns>A list containing this cell edit if the column matches; otherwise, an empty list.</returns>
        public override List<CellEdit> GetEditedCellsInColumn(int indexOfColumn)
        {
            if (indexOfColumn == _columnIndex)
                return new List<CellEdit>(new[] { this });
            return new List<CellEdit>();
        }

        /// <summary>
        /// Returns the edited cell if its row matches the specified index.
        /// </summary>
        /// <param name="indexOfRow">The row index to check.</param>
        /// <returns>A list containing this cell edit if the row matches; otherwise, an empty list.</returns>
        public override List<CellEdit> GetEditedCellsInRow(int indexOfRow)
        {
            if (indexOfRow == _rowIndex)
                return new List<CellEdit>(new[] { this });
            return new List<CellEdit>();
        }

        /// <summary>
        /// Determines whether the edited cell affects the specified row.
        /// </summary>
        /// <param name="indexOfRow">The index of the row to check.</param>
        /// <returns><c>true</c> if this edit affects the specified row; otherwise, <c>false</c>.</returns>
        public override bool ContainsRow(int indexOfRow)
        {
            return _rowIndex == indexOfRow;
        }

        /// <summary>
        /// Adjusts the column index of the edit when a new column is added.
        /// </summary>
        /// <typeparam name="T">The type of the column data (optional)</typeparam>
        /// <param name="indexOfColumn">The index where the column is added. </param>
        /// <param name="columnData">The data of the new column (optional)</param>
        public override void ColumnAdded<T>(int indexOfColumn, T[] columnData = null)
        {
            if (_columnIndex >= indexOfColumn)
                _columnIndex += 1;
            // 
            if (_columnIndex < 0)
            {
                if (_columnIndex * -1 > indexOfColumn)
                {
                    _columnIndex += 1;
                }
                else if (_columnIndex * -1 == indexOfColumn)
                {
                    _columnIndex *= -1;
                }
            }
        }

        /// <summary>
        /// Adjusts the column index of the edit when a column is deleted.
        /// </summary>
        /// <param name="indexOfColumn">The index of the deleted column.</param>
        public override void ColumnDeleted(int indexOfColumn)
        {
            if (_columnIndex < 0)
            {
                if (_columnIndex * -1 >= indexOfColumn)
                    _columnIndex -= 1;
            }
            // 
            if (_columnIndex == indexOfColumn)
                _columnIndex *= -1;
            if (_columnIndex > indexOfColumn)
                _columnIndex -= 1;
        }

        /// <summary>
        /// Adjusts the row index of the edit when a new row is added.
        /// </summary>
        /// <param name="indexOfRow">The index where the row is added.</param>
        /// <param name="rowData">The data of the new row.</param>
        public override void RowAdded(int indexOfRow, object[] rowData = null)
        {
            if (_rowIndex >= indexOfRow)
                _rowIndex += 1;

            if (_rowIndex < 0)
            {
                if (_rowIndex * -1 > indexOfRow)
                {
                    _rowIndex += 1;
                }
                else if (_rowIndex * -1 == indexOfRow)
                {
                    _rowIndex *= -1;
                }
            }
        }

        /// <summary>
        /// Adjusts the row index of the edit when a row is deleted.
        /// </summary>
        /// <param name="indexOfRow">The index of the deleted row.</param>
        public override void RowDeleted(int indexOfRow)
        {
            if (_rowIndex < 0)
            {
                if (_rowIndex * -1 >= indexOfRow)
                    _rowIndex -= 1;
            }

            if (_rowIndex == indexOfRow)
                _rowIndex *= -1;
            if (_rowIndex > indexOfRow)
                _rowIndex -= 1;
        }
        
        /// <summary>
        /// Determines whether the edit contains the specified cell and return its value.
        /// </summary>
        /// <param name="indexOfColumn">The column index to check.</param>
        /// <param name="indexOfRow">The row index to check.</param>
        /// <param name="returnValue">The value of the cell if found.</param>
        /// <returns><c>true</c> if the cell matches the edit; otherwise, <c>false</c>.</returns>
        public override bool ContainsCell(int indexOfColumn, int indexOfRow, ref object returnValue)
        {
            if (indexOfColumn != _columnIndex)
                return false;
            if (indexOfRow != _rowIndex)
                return false;
            returnValue = _editValue;
            return true;
        }

        /// <summary>
        /// Determines whether the edit affects the specified column. 
        /// </summary>
        /// <param name="indexOfColumn">The index of the column to check.</param>
        /// <returns><c>true</c> if the column is affected; otherwise, <c>false</c>.</returns>
        public override bool ContainsColumn(int indexOfColumn)
        {
            return indexOfColumn == _columnIndex;
        }
    }
}

File: src/DataBase_Reader_VB/Table Edits/CellEdit.vb
Public Class CellEdit
    Inherits TableEdit
    Private _rowIndex As Int32
    Private _columnIndex As Int32
    Private ReadOnly _editValue As Object
    Public ReadOnly Property RowIndex As Int32
        Get
            Return _rowIndex
        End Get
    End Property
    Public ReadOnly Property ColumnIndex As Int32
        Get
            Return _columnIndex
        End Get
    End Property
    Public ReadOnly Property Value As Object
        Get
            Return _editValue
        End Get
    End Property
    Public Sub New(indexOfRow As Int32, indexOfColumn As Int32, editValue As Object)
        _rowIndex = indexOfRow
        _columnIndex = indexOfColumn
        _editValue = editValue
    End Sub

    Public Overrides Function GetEditedCellsInColumn(indexOfColumn As Integer) As List(Of CellEdit)
        If indexOfColumn = _columnIndex Then Return New List(Of CellEdit)({Me})
        Return New List(Of CellEdit)
    End Function

    Public Overrides Function GetEditedCellsInRow(indexOfRow As Integer) As List(Of CellEdit)
        If indexOfRow = _rowIndex Then Return New List(Of CellEdit)({Me})
        Return New List(Of CellEdit)
    End Function

    Public Overrides Function ContainsRow(indexOfRow As Integer) As Boolean
        Return _rowIndex = indexOfRow
    End Function

    Public Overrides Sub ColumnAdded(Of T)(indexOfColumn As Integer, Optional ByVal columnData() As T = Nothing)
        If _columnIndex >= indexOfColumn Then _columnIndex += 1
        '
        If _columnIndex < 0 Then
            If _columnIndex * -1 > indexOfColumn Then
                _columnIndex += 1
            ElseIf _columnIndex * -1 = indexOfColumn Then
                _columnIndex *= -1
            End If
        End If
    End Sub

    Public Overrides Sub ColumnDeleted(indexOfColumn As Integer)
        If _columnIndex < 0 Then If _columnIndex * -1 >= indexOfColumn Then _columnIndex -= 1
        '
        If _columnIndex = indexOfColumn Then _columnIndex *= -1
        If _columnIndex > indexOfColumn Then _columnIndex -= 1
    End Sub

    Public Overrides Sub RowAdded(indexOfRow As Integer, Optional ByVal rowData() As Object = Nothing) 
        If _rowIndex >= indexOfRow Then _rowIndex += 1

        If _rowIndex < 0 Then
            If _rowIndex * -1 > indexOfRow Then
                _rowIndex += 1
            ElseIf _rowIndex * -1 = indexOfRow Then
                _rowIndex *= -1
            End If
        End If
    End Sub

    Public Overrides Sub RowDeleted(indexOfRow As Integer)
        If _rowIndex < 0 Then If _rowIndex * -1 >= indexOfRow Then _rowIndex -= 1

        If _rowIndex = indexOfRow Then _rowIndex *= -1
        If _rowIndex > indexOfRow Then _rowIndex -= 1
    End Sub

    Public Overrides Function ContainsCell(indexOfColumn As Integer, indexOfRow As Integer, ByRef returnValue As Object) As Boolean
        If indexOfColumn <> _columnIndex Then Return False
        If indexOfRow <> _rowIndex Then Return False
        returnValue = _editValue
        Return True
    End Function

    Public Overrides Function ContainsColumn(indexOfColumn As Integer) As Boolean
        Return indexOfColumn = _columnIndex
    End Function
End Class
